### ISSUE TITLE
No mobile TLS fingerprint support post Chrome v130

### ISSUE THREAD

**Issue by charliedelta02** [lexiforest/curl_cffi#434]
**Is your feature request related to a problem? Please describe.**
I use curl-cffi for chrome to mimic both Desktop (MacOS, Windows) and Mobile (Android) device TLS fingerprints.
Until v129 for chrome, the TLS fingerprint remained the same both on desktop and mobile devices. Therefore, I used `chrome124` impersonate supported by curl-cffi with desktop and mobile user-agents to mimic desktop and mobile behavior respectively.
With chrome v130, it looks like fingerprint is different from `chrome124` on android mobile devices. All my requests to websites fail with Captcha if chrome v130 is used with mobile user-agents.
It does appear to me that chrome v130 with desktop user-agents (macOS and Windows) still works well at the time of writing this with `chrome124`.

**Describe the solution you'd like**
TLS finderprint support for chrome on mobile devices post v130.

**Describe alternatives you've considered**
A possibility is to use `safari17_2_ios` support with iOS user-agents to mimic an iOS mobile device with the latest sarari versions such as 17_6, 18_0.

**Comment by lexiforest** [comment]
I guess Google is A/B testing new TLS extensions on Android and Windows. I checked again my Chrome 130 on macOS, and it still shares the same fingerprints with previous versions. I will find Android and Windows devices and try again later.

**Comment by charliedelta02** [comment]
> I guess Google is A/B testing new TLS extensions on Android and Windows. I checked again my Chrome 130 on macOS, and it still shares the same fingerprints with previous versions. I will find Android and Windows devices and try again later.

To confirm, until there is a workable fingerprint for Android and Windows for chrome v130, what should our approach be? Continue using Chrome v129 or lower with `chrome124` impersonate? Or should our approach be different?

**Comment by lexiforest** [comment]
You can use the `ja3` and `akamai` parameter to customize the fingerprints.

**Comment by lexiforest** [comment]
`ja3` is subject to change due to so-called TLS extension permutation mechanism, you should focus on the `ja3n` string.

**Comment by charliedelta02** [comment]
> `ja3` is subject to change due to so-called TLS extension permutation mechanism, you should focus on the `ja3n` string.

Understood. Thank you so very much. I will implement this for the time being. One minor clarification though.

From https://tls.browserleaks.com/json, for Chrome v131 on an Android device, I see this as the ja3n output:
```
"user_agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/[131.0.0.0](http://131.0.0.0/) Mobile Safari/537.36"
```
```
"ja3n_hash": "473f0e7c0b6a0f7b049072f4e683068b",
"ja3n_text": "771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0"
```
```
"akamai_hash": "52d84b11737d980aef856699f885ca86",
"akamai_text": "1:65536;2:0;4:6291456;6:262144|15663105|0|m,a,s,p"
```

As per this, my curl-cffi script should be as follows, correct?

```
ja3_fingerprint_str = "771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0"
akamai_str = "1:65536;2:0;4:6291456;6:262144|15663105|0|m,a,s,p"
response = await s.get(
                        url,
                        proxies={'http': proxy, 'https': proxy},
                        headers=headers,
                        timeout=20,
                        ja3=ja3_fingerprint_str,
                        akamai=akamai_str
                    )
```

**Comment by charliedelta02** [comment]
> ```
> ja3_fingerprint_str = "771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0"
> akamai_str = "1:65536;2:0;4:6291456;6:262144|15663105|0|m,a,s,p"
> response = await s.get(
>                        "https://tls.browserleaks.com/json",
>                         proxies={'http': proxy, 'https': proxy},
>                         headers=headers,
>                         timeout=20,
>                         ja3=ja3_fingerprint_str,
>                         akamai=akamai_str
>                     )
> ```

@lexiforest I tested the following script for Chrome v131 for an Android device.
The `ja3n` and `akamai` strings were fetched by accessing https://tls.browserleaks.com/json from an android device on chrome v131.

_**My Observation:**_
_I sent multiple requests to https://tls.browserleaks.com/json with the updated `ja3` and `akamai` parameters as shown above.
Most of the times the `ja3n_text` on https://tls.browserleaks.com/json is as expected, which is:
```
771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0
```
Once in a while it is not the same as what I have passed. I get a `ja3n_text` such as:
```
771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-21-23-27-35-43-45-51-17513-65037-65281,29-23-24,0
```

**I don't believe this should be the expected behavior, should it?_**

**Comment by lexiforest** [comment]
Extension 21 is padding, which will pop up when the length of the packet is not ideal. But I'm not sure it should be considered as part of the JA3 string, normally it should be ignored.

**Comment by charliedelta02** [comment]
> Extension 21 is padding, which will pop up when the length of the packet is not ideal. But I'm not sure it should be considered as part of the JA3 string, normally it should be ignored.

This is a response I observe quite often from https://tls.browserleaks.com/json when a custom `ja3` and `akamai` parameters are used with curl-cffi `response = await s.get(ja3=, akamai=)`. Might be the behavior of https://tls.browserleaks.com/json. Not exactly sure.
But, I am assuming the process I have followed in the script shared earlier to pass `ja3n` string and `akamai` string is exactly how you have suggested right?

**Comment by charliedelta02** [comment]
@lexiforest I performed a round of testing with websites that perform TLS fingerprinting with custom `ja3n` and `akamai` fingerprint passed via curl-cffi for chrome v131 on Android.
I observed an explosion of captchas with most websites.

**Comment by charliedelta02** [comment]
@lexiforest 
Am i using the `ja3` and `akamai` parameters wrong? I am using the `ja3n` text as the `ja3` parameter generated using browserleaks.com from an Android chrome browser on v131.
```
ja3_fingerprint_str = "771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0"
akamai_str = "1:65536;2:0;4:6291456;6:262144|15663105|0|m,a,s,p"
response = await s.get(
                       "https://tls.browserleaks.com/json",
                        proxies={'http': proxy, 'https': proxy},
                        headers=headers,
                        timeout=20,
                        ja3=ja3_fingerprint_str,
                        akamai=akamai_str
                    )
```

**Comment by charliedelta02** [comment]
> I guess Google is A/B testing new TLS extensions on Android and Windows. I checked again my Chrome 130 on macOS, and it still shares the same fingerprints with previous versions. I will find Android and Windows devices and try again later.

If I am not mistaken, the fingerprint change post v130 has also been updated here on tls-client: https://github.com/bogdanfinn/tls-client/blob/master/profiles/contributed_browser_profiles.go ? So the fingerprint needs to be changed right?

Most scrape requests are being detected now. I know you would have your hands full, but would it be a decent expectation to have support for these versions any time soon? Especially for the mobile devices.

**Comment by lexiforest** [comment]
It turns out [Google decided](https://security.googleblog.com/2024/09/a-new-path-for-kyber-on-web.html) to withdraw the Kyber key exchange algorithm, and went with the ML-KEM method. My Chrome 131 finally receives this update. 

The difference in JA3N string is actually quite simple (25497 -> 4588).

```diff
-771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,25497-29-23-24,0
+771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,4588-29-23-24,0
```

I hope I can find some time to follow Google's move and fix the CI here before Christmas and  new year.

**Comment by charliedelta02** [comment]
> The difference in JA3N string is actually quite simple (25497 -> 4588).
> 
> ```diff
> -771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,25497-29-23-24,0
> +771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,4588-29-23-24,0
> ```

If I try to pass this JA3N string as a parameter, I get the following exception.
```
curl_cffi/requests/session.py", line 326, in _set_ja3_options
    curve_name = TLS_EC_CURVES_MAP[curve_id]
```


> I need to update my patched BoringSSL and also curl-impersonate, it's not a trivial task.
> 
> I hope I can find some time to follow Google's move and fix the CI here before Christmas and new year.

Thank you for your efforts and support. Hoping to see this update soon.
Post the fingerprint change performed by google, all the calls I make to websites using `chrome124` impersonate for both Desktop and Mobile UAs (I am mimicking up to chrome v129 at present) are having a **much higher failure rate** with Captchas.
Until, there is an update from you, is there anything we can do to not being detected as a bot and maintain the success rates?

**Comment by lexiforest** [comment]
> Until, there is an update from you, is there anything we can do to avoid being blocked as a bot and maintain the success rates?

Use the safari targets, if it helps.

**Comment by charliedelta02** [comment]
> Use the safari targets, if it helps.

It indeed will help. My script is flexible enough to use safari as well. The problem lies with the version of curl-cffi. Support for `safari18_0` for Desktop and `safari_18_0_ios` for Mobile is available on curl-cffi _v0.8_. Unfortunately, I am unable to install this version using `pip3 install` (either on Mac or on a Linux machine) and the max version I can go up to is curl-cffi _v0.7.3_ which only supports `safari17_0` for desktop and `safari17_2_ios` for Mobile. Is there a way I can upgrade to _v0.8_?

**Comment by lexiforest** [comment]
```
pip install curl-cffi==0.8.0b1
```

**Comment by charliedelta02** [comment]
> pip install curl-cffi==0.8.0b1

I did try installing these versions (0.8.0, 0.8.0b1). Every time I do this, I get the following exception:
```
ERROR: Ignored the following yanked versions: 0.1.5, 0.7.2
ERROR: Could not find a version that satisfies the requirement curl-cffi==0.8.0b1 (from versions: 0.2.1, 0.2.4, 0.2.5, 0.3.0, 0.3.1, 0.3.2, 0.3.7, 0.3.8, 0.4.0, 0.5.0, 0.5.1, 0.5.2, 0.5.3, 0.5.4, 0.5.5, 0.5.6, 0.5.7, 0.5.9b1, 0.5.9b2, 0.5.9b3, 0.5.9b4, 0.5.9b5, 0.5.9b6, 0.5.9, 0.5.10b1, 0.5.10b2, 0.5.10b3, 0.5.10b4, 0.5.10b5, 0.5.10, 0.6.0b2, 0.6.0b4, 0.6.0b7, 0.6.0b9, 0.6.0, 0.6.1, 0.6.2, 0.6.3b1, 0.6.3, 0.6.4, 0.7.0b4, 0.7.0b6, 0.7.0b7, 0.7.0, 0.7.1, 0.7.3b1, 0.7.3)
ERROR: No matching distribution found for curl-cffi==0.8.0b1
```

**Comment by charliedelta02** [comment]
I will keep this issue open until there is support for Chrome v130 and above for Mobile and Desktop devices if it is fine by you.

**Comment by charliedelta02** [comment]
> ```
> pip install curl-cffi==0.8.0b1
> ```

It is wonderful to know that chrome131 support is now live.
I tested Chrome131 with Mac and Windows desktop devices on websites that TLS fingerprint and they work as expected.
But, for Android Mobile devices, using chrome131 impersonate on same websites leads to multiple failures. 

**Looks like the fingerprint is not the same for Mobile and Desktop devices. Can we get chrome131 support for mobile devices as well?**

**Comment by lexiforest** [comment]
The fingerprint of Chrome 131 for Android has also been confirmed.

```diff
-771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,4588-29-23-24,0
+771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0
```

MLKEM is removed on mobile, there is no new extension, you can simply use the JA3 string as parameter before it is built into curl-impersonate.

**Comment by charliedelta02** [comment]
> 
> MLKEM is removed on mobile, there is no new extension, you can simply use the JA3 string as parameter before it is built into curl-impersonate.

Just to confirm, you are suggesting to pass this as part of `requests.get()` `ja3` parameter right? 
What about the `akamai` parameter?


> The fingerprint of Chrome 131 for Android has also been confirmed.
> 
> +771,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-5-10-11-13-16-18-23-27-35-43-45-51-17513-65037-65281,29-23-24,0
> ```

May I know how to obtain fingerprint? https://tls.browserleaks.com/json does not give the same fingerprint.

Will there be no impersonate support for this fingerprint for Android on curl-cffi??

**Comment by lexiforest** [comment]
> you are suggesting to pass this as part of requests.get() ja3 parameter right?

Yes

> What about the akamai parameter?

No need.

If you have different fingerprint, I suggest that you report yours in a new issue, via the fingprint report template, for better tracking.

I got the above fingerprint from my GooglePlay-installed Chrome 131 for Android, I think this should be a reliable source. And it does make sense, the MLKEM curve can be compute-tensive, so Google disables it on mobile.

**Comment by charliedelta02** [comment]
> If you have different fingerprint, I suggest that you report yours in a new issue, via the fingprint report template, for better tracking.

Never mind. I tested this again with https://tls.browserleaks.com/json and, the fingerprint (ja3n) for Android Chrome v131 matches with the string shared by you above.
```
"ja3n_hash": "78d450322efcbf3f5d2d92f9d7769566",
  "ja3n_text": "771,49199-49171-49191-49200-49172-49192-156-47-60-157-53-61-65-132-49195-49324-49326-49161-49187-49196-49325-49327-49162-49188-158-51-103-159-57-107-69-136-4865-4866-255,0-10-11-13-23-43-51,23-29-24-256-257-258,0",
```
Can we expect Chrome131 Mobile impersonate anytime soon?

**Comment by lexiforest** [comment]
Released as `0.8.0b7`.
